### PR TITLE
Replace fmt.Errorf() with errors.Errorf() in the cli

### PR DESCRIPTION
**Pull request by dnephin**
```bash
#!/bin/bash

set -eu

DIR=./cli

eg -t refactorerr.go -v -w $(find $DIR -type d)

git grep -l 'github.com/docker/docker/vendor/github.com/pkg/errors'  | xargs sed -i -e 's/github.com\/docker\/docker\/vendor\/github.com\/pkg\/errors/github.com\/pkg\/errors/'

# Fix imports of "errors" in files that don't have fmt.Errorf()
grep -vxF -f \
    <(git grep -l '"github.com/pkg/errors"' $DIR) \
    <(git grep -l '"errors"' $DIR) | \
xargs sed -i -e 's/"errors"/"github.com\/pkg\/errors"/'

# Remove duplicate imports of "errors"
git grep -l '"errors"' $DIR | xargs sed -i -e '/"errors"/ d'


goimports -w $DIR
```

```go
package P

import (
	"fmt"
	"github.com/pkg/errors"
)

func before(s string) func(string, ...interface{}) error { return fmt.Errorf }
func after(s string) func(string, ...interface{}) error  { return errors.Errorf }

```

### CODE REVIEW

**Comment by aaronlehmann** [comment]
I have a small nit: the `"github.com/pkg/errors"` import is mixed with stdlib imports (where `fmt` used to be), when it should really be below the stdlib imports.

In general I think using `github.com/pkg/errors` is a good idea. I'm a bit skeptical of mass conversion, because there may be places where we print an error in such a way that we get a stack trace, when that isn't an appopriate context to print a stack trace. For the CLI i think the risk is relatively low, so it's probably okay to do this and then keep an eye out for unexpected stack traces in CLI output. I definitely wouldn't want to do this automated conversion in the daemon.

**Comment by thaJeztah** [comment]
Looks like a flaky test?

```
04:00:49 ----------------------------------------------------------------------
04:00:49 FAIL: check_test.go:101: DockerSuite.TearDownTest
04:00:49 
04:00:49 check_test.go:102:
04:00:49     testEnv.Clean(c, dockerBinary)
04:00:49 environment/clean.go:61:
04:00:49     t.Fatalf("error removing containers %v : %v (%s)", containers, result.Error, result.Combined())
04:00:49 ... Error: error removing containers [420b76fe54e8] : exit status 1 (Error response from daemon: removal of container 420b76fe54e8 is already in progress
04:00:49 )
04:00:49 
04:00:49 
04:00:49 ----------------------------------------------------------------------
04:00:49 PANIC: docker_cli_restart_test.go:279: DockerSuite.TestRestartAutoRemoveContainer
04:00:49 
04:00:49 ... Panic: Fixture has panicked (see related PANIC)
```

**Comment by dnephin** [comment]
> I have a small nit: the `github.com/pkg/errors` import is mixed with stdlib imports (where fmt used to be), when it should really be below the stdlib imports.

Do you know of any tool that will handle this sorting? It seems like `goimport` isn't aware of this convention, so it doesn't quite sort it the way we expect. Python has a bunch of tools (ex https://github.com/timothycrosley/isort) to handle this.

> In general I think using github.com/pkg/errors is a good idea. I'm a bit skeptical of mass conversion, because there may be places where we print an error in such a way that we get a stack trace, when that isn't an appopriate context to print a stack trace.

I think it will only print a stack trace is we use `%+v` , right? Luckily we only do this a handful of times, so it should be easy to verify the uses. I see we are using it to print errors in a few places right now.

> I definitely wouldn't want to do this automated conversion in the daemon.

I was hoping to do this for the next step. I think it would be good to print stack traces from the daemon when debug logging is enabled (or maybe it could be enabled by an env variable like `DOCKER_ENABLE_STACKTRACE`).

Right now it looks like we print using `%+v` on any unhandled 500.

**Comment by aaronlehmann** [comment]
> Do you know of any tool that will handle this sorting? It seems like goimport isn't aware of this convention, so it doesn't quite sort it the way we expect. Python has a bunch of tools (ex https://github.com/timothycrosley/isort) to handle this.

Sorry, I'm not aware of a tool for this,.

> I think it will only print a stack trace is we use `%+v` , right? Luckily we only do this a handful of times, so it should be easy to verify the uses. I see we are using it to print errors in a few places right now.

I think so. Good to hear this should be easy to verify.

> I was hoping to do this for the next step. I think it would be good to print stack traces from the daemon when debug logging is enabled (or maybe it could be enabled by an env variable like DOCKER_ENABLE_STACKTRACE).
> 
> Right now it looks like we print using %+v on any unhandled 500.

The `%+v` was added to provide stack traces for some plugin-specific errors. It turned out that there were other places in the code using `pkg/errors` to generate errors, so this initially was spewing stack traces when any endpoint failed, for example if `inspect` couldn't find an object, or if `pull` had an authentication error.

This was limited to code 500 errors to prevent these benign cases from inserting stack traces into the logs. It's an example of how indiscriminately using `%+v` can be an issue.

I'm open to making more use of `pkg/errors`, but we must make sure that we aren't barfing stack traces in the logs unless it's really an internal error where that stack trace is useful for debugging. Hiding them behind an environment variable probably makes sense. We should also keep in mind that there's probably a fair amount of overhead in capturing a stack trace compared to allocating a normal error string, so we should avoid this package in places where that overhead might be a problem (I don't know if there are any, but I'm thinking about internal functions that return an `error` never meant to be bubbled up to the user).

**Comment by tiborvass** [comment]
Ping @dnephin needs rebase

**Comment by dnephin** [comment]
rebased

**Comment by tiborvass** [comment]
LGTM

File: cli/command/cli.go [hunk 1]
```diff
@@ -1,8 +1,8 @@
 package command
 
 import (
-	"errors"
 	"fmt"
+	"github.com/pkg/errors"
```
Comment (aaronlehmann) [inline cli/command/cli.go]: We should fix the weird import positioning before merging this. It can be done manually if necessary.

**Comment by tiborvass** [comment]
I ran `find . -name '*.go' -exec grep -HnEC1 '"github.com/pkg/errors"' {} \; | while read line; do echo "$line" | awk '{print $2}' | egrep '".*"' | egrep -vq '[a-z]+\.[a-z]+' && echo "$line"; done | cut -d- -f1 | sort -u` to find the list of files that needed a manual fix to address https://github.com/docker/docker/pull/31709#pullrequestreview-29339090 and pushed febbe38.

File: cli/compose/schema/bindata.go [hunk 2]
```diff
@@ -10,27 +10,28 @@ package schema
...
 	"time"
+
+	"github.com/pkg/errors"
 )
 
 func bindataRead(data []byte, name string) ([]byte, error) {
 	gz, err := gzip.NewReader(bytes.NewBuffer(data))
 	if err != nil {
-		return nil, fmt.Errorf("Read %q: %v", name, err)
+		return nil, errors.Errorf("Read %q: %v", name, err)
```
Comment (aaronlehmann) [inline cli/compose/schema/bindata.go]: Since this is generated code, I don't think it's correct to replace `fmt` with `errors` in this file.

**Comment by aaronlehmann** [comment]
LGTM

**Comment by tiborvass** [comment]
From https://jenkins.dockerproject.org/job/Docker-PRs-experimental/32168/console:

```
03:01:19 ----------------------------------------------------------------------
03:01:19 FAIL: docker_cli_authz_plugin_v2_test.go:148: DockerAuthzV2Suite.TestAuthZPluginBadManifestFailsDaemonStart
03:01:19 
03:01:19 [dbc9b9ecd5683] waiting for daemon to start
03:01:19 [dbc9b9ecd5683] daemon started
03:01:19 
03:01:19 [dbc9b9ecd5683] exiting daemon
03:01:19 [dbc9b9ecd5683] waiting for daemon to start
03:01:19 [dbc9b9ecd5683] waiting for daemon to start
03:01:19 [dbc9b9ecd5683] exiting daemon
03:01:19 docker_cli_authz_plugin_v2_test.go:158:
03:01:19     // restarting the daemon without requiring the plugin will succeed
03:01:19     s.d.Restart(c)
03:01:19 daemon/daemon.go:392:
03:01:19     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
03:01:19 ... Error: Error while stopping the daemon dbc9b9ecd5683 : exit status 1
03:01:19 
```

**Review by vdemeester — APPROVED** [review]
LGTM 🦁

**Pull request merged**